### PR TITLE
Feature/issue 981 member billing ledger mvp

### PR DESCRIPTION
**Pull request by invrtd**
# Summary
Implements the MVP billing workflows for member, guest, and commercial-flight charges.
Included
- Adds immutable member billing ledger corrections and payments.
- Defaults member logsheet payments to ledger billing.
- Adds guest payment tracking and treasurer remittance workflow.
- Requires explicit payment methods for guest flights before finalization.
- Treats commercial rides as prepaid and excludes them from member ledgers.
- Adds audit history, reversal protections, migrations, and focused tests.
# Issues
- Closes #986
- Closes #988
- Closes #989
- Related to #981
# Verification
- Focused billing and finalization tests pass.
- Django system checks pass.
- Migration checks pass.
- Pre-commit hooks pass.

### CODE REVIEW

**Comment by pietbarber** [comment]
Here’s the gist of PR #1026:

**In a nutshell:** This PR adds the billing-ledger MVP for member, guest, and commercial-flight charges. It introduces immutable ledger behavior for opening balances, a guest-payment workflow that stays pending until treasurer remittance, and UI/API support for exporting ledgers and managing these new billing states.

### Summary

| Item | Details |
|---|---|
| PR | **Feature/issue 981 member billing ledger mvp** |
| Author | `invrtd` |
| State | Open, not draft |
| Scope | 21 files changed |
| Size | +1293 / -35 |
| Commits | 6 |
| Reviews / comments | No reviews, no review comments, no issue comments shown |
| Mergeability | Mergeable, `clean` |
| Risk | Medium-high, because it changes core billing flows and introduces new ledger invariants/migrations |
| Linked issues | Closes **#986**, **#988**, **#989**; related to **#981** |
| CI status | Not available from the data I fetched |

### What the code changes do

#### Billing core changes
- Adds new ledger entry kinds:
  - `guest_payment_pending`
  - `guest_remittance`
- Adds new fields to `LedgerEntry`:
  - `guest_name`
  - `payment_method`
  - `remits` self-link for guest remittance tracking
- Enforces new validation rules:
  - Guest pending entries must be debits and include guest name + payment method.
  - Guest remittances must be credits, must point to a pending guest payment, and must match the full amount.
  - Opening balances are now one-per-ledger.
- Prevents posting multiple opening balances and adds a separate override path that posts an audited adjustment instead of mutating history.

#### New billing services
- `post_guest_payment_pending(...)`
- `remit_guest_payment(...)`
- `override_opening_balance(...)`

These are the main workflow additions:
- guest flight charges stay as a member liability until the treasurer confirms payment
- remittance clears the pending balance in full
- opening balance corrections are handled via a new adjustment entry rather than rewriting the original record

#### UI changes in billing
- Ledger detail page now:
  - shows running balance per entry
  - offers CSV export
  - includes an opening-balance override modal
  - shows a remittance action for pending guest payments
- New CSV export endpoint sanitizes cells to reduce spreadsheet formula injection risk

#### Logsheet / guest-flight workflow
- Adds a `LogsheetGuestPayment` model and migration
- Guest flights must now have:
  - a responsible member
  - a payment method
- Finalizing a logsheet now validates guest payment completeness
- Guest payments are converted into pending billing entries during finalization
- Guest flights are excluded from normal member charge allocation logic

#### Member payment defaulting
- Existing member logsheet payments now default to `account`
- This aligns with the billing assumption that a payment-method selection alone should not create a payment credit

### Tests added or updated
The PR adds tests for:
- guest payment pending/remittance workflow
- treasurer-only remittance
- one opening balance per ledger
- opening balance override behavior
- CSV export and CSV sanitization
- guest flight finalization requiring explicit payment method
- member payment defaulting to `account`

### Notable design choices
- The PR keeps entries immutable and uses new entries for corrections/remittances.
- It explicitly models guest payments as a two-step workflow:
  1. logsheet finalization creates a pending member liability
  2. treasurer confirms full remittance later
- It treats opening balance corrections as audit-preserving adjustments, not edits.

### Potential risk areas
- This is core billing logic, so any validation mismatch could block finalization or ledger operations.
- The guest payment/remittance model is fairly strict; partial remittance is intentionally unsupported.
- The new opening-balance override logic adds complexity around balance calculation and reversal handling.

Absolutely — here are all three.

## 1) File-by-file walkthrough

### `billing/forms.py`
- Extends `ManualEntryForm` so `opening_balance` can be hidden from normal entry creation once one exists.
- Adds `OpeningBalanceOverrideForm` for audited balance corrections.
- Adds `GuestRemittanceForm` for treasurer confirmation of guest-payment collection.

### `billing/migrations/0007_one_opening_balance_per_ledger.py`
- Adds a database constraint so each ledger can have only one `opening_balance`.
- Includes a preflight migration check that fails if existing data already violates that rule.

### `billing/migrations/0008_guest_payment_pending.py`
- Adds new `LedgerEntry` fields for guest-payment workflow:
  - `guest_name`
  - `payment_method`
  - `remits`
- Updates validation constraints so guest payment pending/remittance entries are legal.
- Adds a remittance integrity constraint.

### `billing/migrations/0009_alter_ledgerentry_kind.py`
- Adds the new ledger kinds:
  - `guest_payment_pending`
  - `guest_remittance`

### `billing/models.py`
- Mirrors the migration changes in the model:
  - new entry kinds
  - new fields
  - one-opening-balance-per-ledger constraint
  - guest remittance validation
- Adds model-level `clean()` rules for guest payment and remittance consistency.

### `billing/services.py`
- Adds the new business operations:
  - `post_guest_payment_pending`
  - `remit_guest_payment`
  - `override_opening_balance`
- Tightens `post_opening_balance()` so only one opening balance can exist.
- Adds guardrails for reversing guest payments.
- Makes entry posting aware of guest payment metadata.

### `billing/templates/billing/ledger_detail.html`
- Adds:
  - running balance column
  - CSV export button
  - opening-balance override modal
  - “Confirm full remittance” action for pending guest payments
- Changes ledger rendering from simple entry rows to statement-style rows.

### `billing/tests/test_manual_transactions.py`
- Covers:
  - guest payment pending/remittance flow
  - treasurer-only remittance
  - one opening balance only
  - opening balance override behavior
- Also checks reversal edge cases around remitted guest payments.

### `billing/tests/test_views.py`
- Tests:
  - guest payment remittance endpoint
  - CSV export and CSV sanitization
  - opening balance override view
  - access control for treasurers vs members

### `billing/urls.py`
- Adds routes for:
  - ledger CSV export
  - opening balance override
  - guest payment remittance

### `billing/views.py`
- Updates ledger detail to render statement rows and override state.
- Adds CSV export endpoint.
- Adds opening balance override POST endpoint.
- Adds guest payment remittance POST endpoint.

### `docs/billing-assumptions.md`
- New policy doc explaining the billing rules the PR implements.
- Useful as the “source of truth” for why the workflow works this way.

### `logsheet/migrations/0031_logsheetpayment_default_account.py`
- Makes member payment method default to `account`.

### `logsheet/migrations/0032_logsheetguestpayment.py`
- Adds the new `LogsheetGuestPayment` model to store guest-flight settlement metadata.

### `logsheet/models.py`
- Adds `LogsheetGuestPayment`.
- Sets `LogsheetPayment.payment_method` default to `account`.

### `logsheet/services.py`
- Finalization now validates guest payment completeness.
- Posts guest payment pending ledger entries during finalization.
- Uses locking more carefully around flights.

### `logsheet/templates/logsheet/manage_logsheet_finances.html`
- Adds a guest-flight payments section to the finances UI.
- Lets treasurers choose responsible member, payment method, and note per guest flight.

### `logsheet/tests/test_billing_finalization.py`
- Verifies guest flight finalization posts pending billing entries.
- Verifies guest finalization fails without explicit payment method.

### `logsheet/tests/test_finances_ui_and_split.py`
- Verifies member payment defaults to `account`.
- Confirms the UI reflects that default.

### `logsheet/utils/flight_charges.py`
- Treats guest flights like commercial rides for charge allocation purposes.
- Prevents normal member allocation logic from applying to guest flights.

### `logsheet/views.py`
- Excludes guest flights from normal member finance calculations.
- Builds guest payment rows for the UI.
- Saves guest payment choices from the finance page.
- Blocks finalization if any guest flight lacks a responsible member or payment method.

---

## 2) Reviewer-style risk assessment

**Overall risk: Medium-high**

### Why it’s risky
- This touches core billing and finalization flows, which are usually high-blast-radius areas.
- It introduces new invariants at both the model and database levels.
- It changes how ledger history behaves, especially opening balances and guest remittances.
- It adds a new two-step workflow that can block logsheet finalization if incomplete.

### Strengths
- Good test coverage for the new paths.
- Business rules are documented in `docs/billing-assumptions.md`.
- The implementation is mostly additive rather than rewriting existing flows.
- Uses explicit DB constraints and service-layer validation, which is good for correctness.
- The opening balance override preserves audit history instead of mutating old entries.

### Risks / things I’d review closely
- **Constraint complexity:** the new `CheckConstraint`/`UniqueConstraint` logic on `LedgerEntry` is fairly nuanced.
- **Remittance correctness:** guest remittance must match exactly and only once; this is intentionally strict, but it may be a source of user friction or edge cases.
- **Finalization coupling:** guest payments are now a required precondition for finalization, so incomplete UI data can block workflows.
- **Balance calculations:** the opening balance override logic is a bit tricky because it has to account for reversals and prior overrides.
- **Template/data flow complexity:** the finances page now mixes member settlement and guest settlement logic, which increases maintenance burden.

### Merge-readiness view
- If the tests are passing, this looks reasonably structured and internally consistent.
- I’d still want to verify:
  - migration ordering
  - the DB constraints behave correctly on existing data
  - the new guest-payment UI path is understandable to treasurers
  - remittance and reversal edge cases don’t create dead-end states

---

## 3) UI vs backend breakdown

### UI changes
- Ledger detail page:
  - adds running balance
  - adds CSV export
  - adds opening-balance override modal
  - adds remittance action for guest pending payments
- Logsheet finance page:
  - adds a dedicated guest-flight payments section
  - lets users assign responsible member, payment method, and note
  - blocks finalization if guest payment info is incomplete
- New docs:
  - billing assumptions/runbook-style guidance

### Backend changes
- Billing model:
  - new entry kinds
  - new fields for guest payment metadata
  - stricter ledger constraints
- Billing services:
  - guest pending posting
  - guest remittance confirmation
  - opening balance override adjustment
- Logsheet finalization:
  - now creates guest pending billing entries
  - validates required guest payment data
- Data migrations:
  - adds persistent support for guest payments
  - enforces one opening balance per ledger

### Practical effect
- **Before:** guest flights and opening balances were less strictly modeled.
- **After:** guest flights become explicit pending liabilities until treasurer confirmation, and opening balances become immutable with audited corrections instead of edits.

**Comment by pietbarber** [comment]
Documentation Feedback: 
1. we need to have ERDs updated for the `billing/docs/models.md` file. 
2. the ERDs for `logsheets/docs/models.md` probably needs to be updated, too.

**Comment by pietbarber** [comment]
Both ERDs are now updated for this PR:

1. `billing/docs/models.md` — `LedgerEntry` now shows the new `guest_name`, `payment_method`, and `remits` (self-FK) fields, plus the `LedgerEntry o|--o| LedgerEntry : "remits"` relationship. The Kind table adds `guest_payment_pending` (debit) and `guest_remittance` (credit, must `remits` a same-ledger collection), and the constraints section documents `billing_guest_remittance_has_collection` and `billing_one_opening_balance_per_ledger`.
2. `logsheet/docs/models.md` — added the new `LogsheetGuestPayment` model (per-flight guest settlement + responsible member, OneToOne to `Flight`) with its relationships; corrected `LogsheetPayment.payment_method` to default `"account"` (on-account) per migration `0031`, and removed the `amount` field the model doesn't actually have.

Committed on top of the current PR head (`59af29d4`) as `9dd24f36` — pushed to the PR head branch.

File: logsheet/services.py [hunk 1]
```diff
@@ -49,6 +52,47 @@ def finalize_logsheet_financials(
...
+        guest_payments = locked_logsheet.guest_payments.select_related(
+            "flight", "responsible_member"
+        )
+        guest_flight_ids = {
+            flight.pk
+            for flight in locked_flights
+            if (flight.guest_pilot_name or "").strip()
+        }
+        recorded_guest_flight_ids = {payment.flight_id for payment in guest_payments}
+        missing_guest_payments = guest_flight_ids - recorded_guest_flight_ids
```
[the remark this answers is not in the document]
Comment (pietbarber) [reply logsheet/services.py]: Fixed all three sub-points in `finalize_logsheet_financials`. (1) The required-guest set now excludes commercial rides (`... and not flight.commercial_ride`) and the posting loop skips any guest row whose flight is a commercial ride, so rides are not required/posted. (2) The guest payments are materialized to a single list and reused for both the set comprehension and the posting loop, so the queryset is no longer evaluated twice. (3) The rows are now locked alongside the logsheet/flights: we lock the base table first (`guest_payments.select_for_update().values_list("pk", flat=True)`, no joins, since PostgreSQL rejects FOR UPDATE on the nullable side of an outer join and `responsible_member` is nullable) and then re-fetch with `select_related("flight", "responsible_member")` for the posting loop. Verified: billing + logsheet finalization/finances test suites pass.

File: logsheet/views.py [hunk 2]
```diff
@@ -2684,7 +2685,13 @@ def manage_logsheet_finances(request, pk):
...
     flights = logsheet.flights.select_related(
         "pilot", "instructor", "glider", "towplane", "split_with"
-    ).exclude(commercial_ride=True)
+    ).exclude(
+        Q(commercial_ride=True)
+        | (Q(guest_pilot_name__isnull=False) & ~Q(guest_pilot_name=""))
+    )
+    guest_flights = logsheet.flights.select_related("pilot").filter(
+        Q(guest_pilot_name__isnull=False) & ~Q(guest_pilot_name="")
+    )
```
[the remark this answers is not in the document]
Comment (pietbarber) [reply logsheet/views.py]: Fixed. `guest_flights` in `manage_logsheet_finances` now filters `~Q(commercial_ride=True)`, so a commercial ride carrying a `guest_pilot_name` no longer gets a `LogsheetGuestPayment` row (and therefore no pending guest-ledger entry). This matches the stated rule that commercial rides are prepaid and handled through the commercial-ride ticket flow. The same exclusion was also applied to the required-guest set and posting loop in `finalize_logsheet_financials` so commercial rides are not double-posted (see the services.py thread reply).

File: logsheet/views.py [hunk 3]
```diff
@@ -2740,6 +2747,22 @@ def flight_costs(f):
...
+            defaults={
+                "guest_name": guest_flight.guest_pilot_name.strip(),
+                "amount": quantize_currency(flight_costs(guest_flight)["total"]),
+                "responsible_member": guest_flight.pilot,
+            },
+        )
+        if created and guest_payment.amount <= 0:
+            guest_payment.delete()
+            continue
+        guest_payment_data.append(guest_payment)
```
[the remark this answers is not in the document]
Comment (pietbarber) [reply logsheet/views.py]: Fixed both points. The guest-payment loop now recomputes `computed_total = quantize_currency(flight_costs(flight)["total"])` on every load and, for an already-existing row, re-syncs `guest_name` and `amount` (`save(update_fields=["guest_name", "amount"])`) so a stale amount is never carried into finalization — while preserving the user-edited `responsible_member`/`payment_method`/`note`. The non-positive guard is no longer gated on `created`: a fresh or pre-existing row whose computed total is `<= 0` is deleted, since `LedgerEntry` enforces `amount > 0` and such a row would fail finalization.

File: logsheet/views.py [hunk 4]
```diff
@@ -2868,6 +2905,22 @@ def flight_costs(f):
...
+                guest_payment.responsible_member_id = responsible_member_id
+            guest_payment.payment_method = (
+                request.POST.get(f"guest_payment_method_{guest_payment.pk}") or None
+            )
+            guest_payment.note = request.POST.get(
+                f"guest_payment_note_{guest_payment.pk}", ""
+            ).strip()
+            guest_payment.save(
+                update_fields=["responsible_member", "payment_method", "note"]
+            )
```
[the remark this answers is not in the document]
Comment (pietbarber) [reply logsheet/views.py]: Fixed. The guest-payment POST save loop is now wrapped in `if not logsheet.finalized:`, so a crafted POST cannot alter `responsible_member`/`payment_method`/`note` after the ledger entries have been posted. The `finalize` branch is unaffected (it still runs after this guard, but only fires on first-time finalize when the guard was also true — i.e., not yet finalized).

File: logsheet/views.py [hunk 5]
```diff
@@ -2740,6 +2751,32 @@ def flight_costs(f):
...
+            # flight costs so a stale amount is never carried into finalization.
+            guest_payment.guest_name = guest_flight.guest_pilot_name.strip()
+            guest_payment.amount = computed_total
+            guest_payment.save(update_fields=["guest_name", "amount"])
+        # A non-positive total (fresh or pre-existing) cannot be posted to the
+        # ledger (LedgerEntry enforces amount > 0), so drop it.
+        if guest_payment.amount <= 0:
+            guest_payment.delete()
+            continue
+        guest_payment_data.append(guest_payment)
```
[the remark this answers is not in the document]
Comment (pietbarber) [reply logsheet/views.py]: Fixed. `manage_logsheet_finances` now branches on `logsheet.finalized`: when the logsheet is finalized, the view no longer calls `get_or_create`, re-syncs, or deletes any `LogsheetGuestPayment` rows. Instead it loads the existing rows read-only (scoped to the same `guest_flights` the non-finalized path would show, with `select_related("flight", "responsible_member")`) for display. The template already disables the guest inputs when finalized (lines 314/326/337), so the frozen state is consistent between server and UI.

File: logsheet/views.py [hunk 6]
```diff
@@ -2868,6 +2921,25 @@ def flight_costs(f):
...
+                    guest_payment.responsible_member_id = responsible_member_id
+                guest_payment.payment_method = (
+                    request.POST.get(f"guest_payment_method_{guest_payment.pk}") or None
+                )
+                guest_payment.note = request.POST.get(
+                    f"guest_payment_note_{guest_payment.pk}", ""
+                ).strip()
+                guest_payment.save(
+                    update_fields=["responsible_member", "payment_method", "note"]
+                )
```
[the remark this answers is not in the document]
Comment (pietbarber) [reply logsheet/views.py]: Fixed in `7fa46a0c`. The guest-payment POST handler now resolves the selected responsible member safely: a blank selection clears it to `None` (so the finalize "missing responsible member" check now works correctly), a valid member pk is assigned, and a non-integer or non-member value is ignored (preserving the prior value) instead of being written raw — preventing the FK-violation 500 and the check bypass. Added regression tests in `test_finances_ui_and_split.py` for the clear, assign, and ignore-invalid paths.

File: logsheet/services.py [hunk 7]
```diff
@@ -49,6 +54,79 @@ def finalize_logsheet_financials(
...
+        # Lock the guest-payment rows alongside the logsheet/flights so their
+        # details cannot change mid-transaction. Lock the base table only (no
+        # joins): PostgreSQL rejects FOR UPDATE on the nullable side of an outer
+        # join, and responsible_member is nullable. Then re-fetch the related
+        # fields for the posting loop below (materialized to a list once).
+        locked_guest_pks = list(
+            locked_logsheet.guest_payments.select_for_update().values_list(
+                "pk", flat=True
+            )
+        )
```
[the remark this answers is not in the document]
Comment (pietbarber) [reply logsheet/services.py]: Fixed in `d134259f`. In `logsheet/services.py`, guest-payment settlement validation is now un-gated from `billing_enabled`:

- The guest-payment lock + completeness validation (row presence for payable guest flights, responsible member, payment method, positive amount) now runs unconditionally before the finalize flag is set, so a logsheet can no longer be finalized while billing is disabled without complete guest settlement rows.
- `post_guest_payment_pending` calls still only run when billing is enabled, so no ledger entries are posted while billing is off.
- Added regression test `test_guest_payment_completeness_required_even_when_billing_disabled` (logsheet/tests/test_billing_finalization.py) covering both the rejection path (missing settlement row with billing disabled) and the successful finalization path (complete row, no ledger entries posted).

File: logsheet/services.py [hunk 8]
```diff
@@ -49,6 +54,85 @@ def finalize_logsheet_financials(
...
+        if not guest_payment.responsible_member_id:
+            raise ValidationError(
+                "Every guest payment must identify a responsible member."
+            )
+        if not guest_payment.payment_method:
+            raise ValidationError("Every guest payment must identify a payment method.")
+        if guest_payment.amount <= 0:
+            raise ValidationError(
+                "Every payable guest flight must have a positive payment amount."
+            )
```
[the remark this answers is not in the document]
Comment (pietbarber) [reply logsheet/services.py:139]: Fixed. `finalize_logsheet_financials` now builds `payable_guest_frozen_totals` from the frozen `*_actual` fields and syncs each payable guest row to that authoritative total (saving with `update_fields=["amount"]`) before validation and before `post_guest_payment_pending`. The pending entry is therefore always posted from the frozen total, never a stale row amount. Added `test_guest_finalization_uses_frozen_total_not_stale_row_amount` and updated the non-positive-amount test, since a payable flight's zero-amount row is now healed to the frozen total and posted rather than rejected.

File: logsheet/views.py [hunk 9]
```diff
@@ -2779,7 +2841,8 @@ def flight_costs(f):
             "total": Decimal("0.00"),
         }
     )
-    for flight, costs in flight_data:
+    for flight in flights:
+        costs = flight_costs(flight)
         tow = costs["tow"] or Decimal("0.00")
```
[the remark this answers is not in the document]
Comment (pietbarber) [reply logsheet/views.py:2864]: Fixed. Added a `costs_by_flight` lookup derived from `flight_data` and used it in both `pilot_summary` and `member_charges`, so `flight_costs` is computed exactly once per flight and all summaries reuse the identical cached values.

File: billing/services.py [hunk 10]
```diff
@@ -331,6 +380,65 @@ def post_manual_charge(
...
+
+@transaction.atomic
+def remit_guest_payment(*, entry, actor, effective_date, reference=""):
+    """Clear a guest payment in full after treasurer confirmation."""
+    require_manual_transaction_access(actor)
+    original = LedgerEntry.objects.select_for_update().get(pk=entry.pk)
+    if original.kind != LedgerEntry.Kind.GUEST_PAYMENT_PENDING:
+        raise ValidationError("Only pending guest payments can be remitted.")
+    if hasattr(original, "remittance"):
+        raise ValidationError("This guest payment has already been remitted.")
```
[the remark this answers is not in the document]
Comment (pietbarber) [reply billing/services.py:429]: Fixed. `remit_guest_payment` now raises `ValidationError("This guest payment was reversed and can no longer be remitted.")` when the entry already has a reversal, before the existing `remittance` guard. A reversed collection no longer owes anything, so remitting it would have recorded a credit against a liability that no longer exists. Added a regression test in `test_manual_transactions.py` covering the remittance-reversed → collection-reversed → remittance-blocked workflow.

File: billing/forms.py [hunk 11]
```diff
@@ -36,5 +43,29 @@ def clean(self):
...
+        min_value=Decimal("0.01"), max_digits=12, decimal_places=2
+    )
+    effect = forms.ChoiceField(choices=EFFECT_CHOICES)
+    effective_date = forms.DateField(
+        initial=date.today, widget=forms.DateInput(attrs={"type": "date"})
+    )
+    # Keep below 255 so the "Opening balance override: " prefix added at
+    # posting time still fits LedgerEntry.member_description (max 255).
+    description = forms.CharField(max_length=233)
+    reason = forms.CharField(widget=forms.Textarea(attrs={"rows": 3}))
```
[the remark this answers is not in the document]
Comment (pietbarber) [reply billing/forms.py]: Fixed in 3aa605ef. Both limits now derive from a single shared constant, so they can't drift: `billing/services.py` defines `OPENING_BALANCE_OVERRIDE_MAX_DESCRIPTION` (= `LedgerEntry.member_description` max_length 255 minus the 22-char "Opening balance override: " prefix = 233), `override_opening_balance()` enforces it, and `OpeningBalanceOverrideForm.description` now sets `max_length=OPENING_BALANCE_OVERRIDE_MAX_DESCRIPTION`. A 230–233-char input passes the form and is accepted by the service (prefix + 233 = 255, fits the column), while 234+ chars are rejected by both layers with the same message. Also in this commit: `post_opening_balance()` now posts with a deterministic `opening-balance:{ledger.pk}` source key, so a double-submit race that slips past the pre-check converges on the winner's row via the `source_key` idempotency path instead of raising an IntegrityError (500). `billing/tests/` 53 passed.

File: billing/migrations/0008_guest_payment_pending.py [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+            field=models.CharField(
+                blank=True,
+                choices=[
+                    ("cash", "Cash"),
+                    ("check", "Check"),
+                    ("zelle", "Zelle"),
+                ],
+                max_length=10,
+            ),
+        ),
```
[the remark this answers is not in the document]
Comment (pietbarber) [reply billing/migrations/0008_guest_payment_pending.py:31]: Good point — I made the backfill explicit by adding `default=""` to both `AddField` operations. Because both fields are `blank=True, max_length=...` CharFields (not `null=True`), adding them to a table that already has rows requires a value for existing rows. Django's schema editor backfills these NOT NULL `blank=True` char fields with `''` automatically (see `BaseDatabaseSchemaEditor._effective_default`), so the migration already works — but making it explicit in the migration file is clearer and removes the reliance on that implicit behavior.

File: logsheet/views.py [hunk 13]
```diff
@@ -1853,11 +1854,25 @@ def manage_logsheet(request, pk):
...
+            # The service performs final completeness checks (e.g. a payable
+            # guest flight with no settlement row) that this view cannot
+            # fully pre-check. Surface those as a user error instead of a
+            # 500.
+            transaction.set_rollback(True)
+            messages.error(
+                request,
+                f"Cannot finalize. {get_validation_message(exc)}",
+            )
+            return redirect("logsheet:manage", pk=logsheet.pk)
```
[the remark this answers is not in the document]
Comment (pietbarber) [reply logsheet/views.py:1872]: Good catch — I removed the `set_rollback(True)` calls. `finalize_logsheet_financials` is itself decorated with `@transaction.atomic`, so a `ValidationError` raised inside it already rolls back any writes it made (the exception propagates out of the atomic block and Django rolls back). Calling `set_rollback(True)` in the view was therefore redundant, and since those call sites are not inside their own atomic block it could even raise `TransactionManagementError`. The view now relies on the service's own atomic block for rollback and just surfaces the validation message to the user.

File: billing/migrations/0008_guest_payment_pending.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (pietbarber) [reply billing/migrations/0008_guest_payment_pending.py:31]: Correction to my previous note on this thread: I did **not** add `default=""` to the migration. I initially tried it, but that creates migration state drift — the model fields (`guest_name`, `payment_method`) deliberately carry **no** default, so `makemigrations --check` then wants to generate an extra `AlterField` migration to strip the default again.

The migration as written is correct and safe as-is: for a NOT NULL `blank=True` CharField, Django's `BaseDatabaseSchemaEditor._effective_default` resolves the implicit backfill to `''` and emits `DEFAULT ''` in the `ALTER TABLE ... ADD COLUMN` statement, so existing `ledgerentry` rows backfill cleanly on `migrate`. That's the intended behavior for `blank=True` string columns, and it keeps the model state and migration state in sync. So no change is needed here — the reviewer's concern is already handled by Django's schema editor.
